src/main/java/picard/illumina/parser/readers/BclReader.java

```diff
@@ -225,34 +225,50 @@ public void remove() {
 
     void advance() {
         int totalCycleCount = 0;
-        final BclData data = new BclData(outputLengths);
-        for (int read = 0; read < outputLengths.length; read++) {
-            for (int cycle = 0; cycle < outputLengths[read]; ++cycle) {
-                try {
 
-                    final int readByte;
-                    try {
-                        readByte = this.streams[totalCycleCount].read();
-                    } catch (IOException e) {
-                        // when logging the error, increment cycle by 1, since totalCycleCount is zero-indexed but Illumina directories are 1-indexed.
-                        throw new IOException(String.format("Error while reading from BCL file for cycle %d. Offending file on disk is %s",
-                                (totalCycleCount+1), this.streamFiles[totalCycleCount].getAbsolutePath()), e);
-                    }
+        try {
+            // See how many clusters we can read and then make BclData objects for them
+            final int clustersRead   = this.streams[0].read(buffer);
+            if (clustersRead == -1) return;
 
-                    if (readByte == -1) {
-                        queue = null;
-                        return;
-                    }
+            final BclData[] bclDatas = new BclData[clustersRead];
+            for (int i=0; i<clustersRead; ++i) {
+                bclDatas[i] = new BclData(outputLengths);
+            }
 
-                    decodeBasecall(data, read, cycle, readByte);
-                    totalCycleCount++;
-                } catch (final IOException ioe) {
-                    throw new RuntimeIOException(ioe);
-                }
+            // Process the data from the first cycle since we had to read it to know how many clusters we'd get
+            updateClusterBclDatas(bclDatas, 0, 0);
+            totalCycleCount += 1;
+
+            for (int read = 0; read < numReads; ++read) {
+                final int readLen = this.outputLengths[read];
+                final int firstCycle = (read == 0) ? 1 : 0;  // For the first read we already did the first cycle above
+
+                for (int cycle = firstCycle; cycle < readLen; ++cycle) {
+                    final int n = this.streams[totalCycleCount].read(buffer, 0, clustersRead);
+                    assert n == clustersRead;
+                    totalCycleCount += 1;
 
+                    updateClusterBclDatas(bclDatas, read, cycle);
+                }
             }
+
+            for (final BclData data : bclDatas) this.queue.add(data);
```

Comment:
Can you split this up with a '{'
```suggestion
            for (final BclData data : bclDatas) {
              this.queue.add(data);
            }
```